rfc/20250721-new-plugin-protocol.md

```diff
@@ -0,0 +1,361 @@
+
+**Scope of this RFC**: This document focuses exclusively on:
+- The wire protocol design using msgpack over stdio
+- Language-agnostic plugin communication patterns
+- Type serialization for OpenTofu's cty type system
+- Protocol extensibility for future capabilities
+
+**Out of scope** (to be addressed in separate RFCs):
+- Plugin discovery and installation mechanisms
+- Registry integration and distribution
+- Security model and sandboxing
+- Specific implementations of middleware, state storage, or provider plugins
+- Migration strategies from existing protocols
+
+The goal is to establish a solid foundation for language-agnostic plugin development that can support various plugin types while maintaining OpenTofu's type safety and feature richness.
+
+## Motivation
+
+The current OpenTofu plugin ecosystem has a critical bottleneck: **the Terraform Plugin SDK/Framework is the only actively maintained path to building extensions, and it's locked to Go**.
+
+While the underlying protocol uses gRPC and protobuf, virtually no one interacts with these directly. The reality is:
+
+1. **Go-only ecosystem**: The Plugin SDK/Framework requires Go expertise, excluding developers who work in other languages
+2. **Heavyweight for simple tasks**: Adding a simple provider-defined function? A full provider is required. The SDK wasn't designed for these lightweight use cases.
+
+### The Cost of Go Lock-in
+
+By limiting extensions to Go, we're excluding teams that work in Python, TypeScript, Rust, Ruby, Java, and many other languages. 
+
+Each team has their preferred tooling and expertise - and I don't believe that we should force them to learn Go just to extend OpenTofu.
+
+These teams currently resort to:
+- Parsing plan JSON files (missing critical runtime context)
+- Writing wrapper scripts around `tofu` commands and reading either the human-readable or machine-readable output
+- Building complex CI/CD pipelines for what should be simple validations
+- Maintaining separate tools that duplicate OpenTofu's resource understanding
+
+### Why Language Agnosticism Matters
+
+The proposed protocol (codename "Seasonings") addresses this by:
+- Using stdio communication - available in every programming language
+- Adopting msgpack for efficient, typed serialization
+- Providing a simple message-based protocol that's easy to implement
+- Supporting different capability levels - from simple hooks to full providers
+- Maintaining full compatibility with OpenTofu's rich type system
+
+This isn't about replacing the Plugin SDK - it's about opening the door to extensions that the current system wasn't designed to support.
+
+## Detailed Design
+
+### Protocol Overview
+
+The Seasonings protocol uses [msgpack-rpc](https://github.com/msgpack-rpc/msgpack-rpc/blob/master/spec.md) communicated over standard input/output (stdio) with message framing.
```

Comment:
So we'd be using this standard, but with newer libraries?  Most of the libraries that are linked via that spec have not been updated in years.  I like that the spec is unchanged, but worry about the libraries.

I do see that https://github.com/msgpack has bindings that are kept up to date.

I'm assuming that we'd  recommend people implement the (very simple) spec using more up-to-date libraries?

Reply:
I think we have 2 options:

- We can implement our own rpc style bindings across msgpack and use the up-to-date libs, or we could use msgpack-rpc directly and then get something out of the box. I didn't consider the maintainability of these packages and maybe we could implement our selves quite easily for low-maintainance.

Reply:
For better or worse we already have `github.com/vmihailenco/msgpack/v5` as an indirect dependency through `go-cty`, so if we do end up writing a msgpack-rpc implementation ourselves (which doesn't _seem_ like it would be a lot of work, but famous last words :upside_down_face:) then we could use that same library to avoid taking on another dependency.

(We do unfortunately also currently have `github.com/hashicorp/go-msgpack` in play as an indirect dependency through the Consul API module for the HashiCorp Consul state storage implementation, but hopefully we'll get to shed _that_ dependency once we deal with state storage via plugins rather than built-in.)